### PR TITLE
impl(GCS+gRPC): optimize reads with Cords enabled

### PR DESCRIPTION
**Pull request by coryan**
The previous version of the code was making an extra copy of the data if `google.storage.v2.ChecksummedData` used the `[ctype = CORD]` annotation *and* the annotation was supported by Protobuf. I expect this to be the common (but not only) case in the near future, it is worth reducing the data copy.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11166)

### CODE REVIEW

File: google/cloud/storage/internal/grpc_buffer_read_object_data.h [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+namespace cloud {
+namespace storage_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+/**
+ * Buffer data received from `ReadObject()`.
+ *
+ * The client library API to download objects keeps the download open until all
+ * the data is received (or the download is interrupted by the application).
+ * While downloading data, the application requests fixed amounts of data. This
```
Comment (devbww) [inline google/cloud/storage/internal/grpc_buffer_read_object_data.h:33]: s/ downloading data/ downloading/ perhaps?

Comment (devbww) [inline google/cloud/storage/internal/grpc_buffer_read_object_data.h:33]: s/data. This/data, which/

File: google/cloud/storage/internal/grpc_buffer_read_object_data.h [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+ * may be smaller than the amount of data received in a
+ * `google.storage.v2.ReadObjectResponse` message. This class buffers any
+ * excess data until the application requests more.
+ *
+ * It may be possible to further optimize this class by using `std::string` and
+ * `absl::string_view` if `g.c.s.v2.ChecksummedData` does not use `absl::Cord`.
+ * We expect that case to be rare in the near future, and while it would save
+ * some allocations, but will not save any data copies. The data copies are
```
Comment (devbww) [inline google/cloud/storage/internal/grpc_buffer_read_object_data.h:41]: s/but/it/

File: google/cloud/storage/internal/grpc_buffer_read_object_data.h [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+ */
+class GrpcBufferReadObjectData {
+ public:
+  GrpcBufferReadObjectData() = default;
+
+  /// Fill @p buffer from the internal buffers.
+  std::size_t FillBuffer(char* buffer, std::size_t n);
+
+  /**
+   *  Save @p contents in the internal buffers and use them to fill @p buffer.
```
Comment (devbww) [inline google/cloud/storage/internal/grpc_buffer_read_object_data.h:52]: s/ Save/Save/

File: google/cloud/storage/internal/grpc_buffer_read_object_data.h [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+   *
+   * This overload is used when Protobuf does **not** implement `[ctype = CORD]`
+   * or if the storage service proto files lacks this annotation.
```
Comment (devbww) [inline google/cloud/storage/internal/grpc_buffer_read_object_data.h:55]: s/lacks/lack/

File: google/cloud/storage/internal/grpc_buffer_read_object_data.h [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+   */
+  std::size_t HandleResponse(char* buffer, std::size_t n, std::string contents);
+
+  /**
+   *  Save @p contents in the internal buffers and use them to fill @p buffer.
```
Comment (devbww) [inline google/cloud/storage/internal/grpc_buffer_read_object_data.h:60]: s/ Save/Save/

File: google/cloud/storage/internal/grpc_buffer_read_object_data.cc [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+#include <algorithm>
+
+namespace google {
+namespace cloud {
+namespace storage_internal {
+
+std::size_t GrpcBufferReadObjectData::FillBuffer(char* buffer, std::size_t n) {
+  std::size_t offset = 0;
+  for (auto v : spill_view_.Chunks()) {
+    if (offset >= n) break;
```
Comment (devbww) [inline google/cloud/storage/internal/grpc_buffer_read_object_data.cc:25]: Opt: I thought you were on the weakest-loop-guard team ... so `if (offset == n) break`?

File: google/cloud/storage/internal/grpc_buffer_read_object_data.cc [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+  spill_view_ = spill_view_.Subcord(offset, spill_view_.size() - offset);
+  return offset;
+}
+
+std::size_t GrpcBufferReadObjectData::HandleResponse(char* buffer,
+                                                     std::size_t n,
+                                                     std::string contents) {
+  contents_ = absl::Cord(std::move(contents));
+  spill_view_ = contents_;
+  return FillBuffer(buffer, n);
```
Comment (devbww) [inline google/cloud/storage/internal/grpc_buffer_read_object_data.cc:39]: Perhaps just `return HandleResponse(absl::Cord(std::move(contents)))`?

File: google/cloud/storage/internal/grpc_buffer_read_object_data.h [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_INTERNAL_GRPC_BUFFER_READ_OBJECT_DATA_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_INTERNAL_GRPC_BUFFER_READ_OBJECT_DATA_H
+
+#include "google/cloud/version.h"
+#include "absl/strings/cord.h"
+#include "absl/strings/string_view.h"
```
Comment (devbww) [inline google/cloud/storage/internal/grpc_buffer_read_object_data.h:20]: Unused?

File: google/cloud/storage/internal/grpc_buffer_read_object_data_test.cc [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    actual.append(buf, buf + n);
+    return n;
+  };
+  auto fill = [&] {
+    auto n = buffer.FillBuffer(buf, kBufferSize);
+    actual.append(buf, buf + n);
+    return n;
+  };
+  for (auto n = response(); n == kBufferSize; n = fill())
+    ;
```
Comment (devbww) [inline google/cloud/storage/internal/grpc_buffer_read_object_data_test.cc:42]: Opt: `s/;/continue;/`, or `{}`?

File: google/cloud/storage/internal/grpc_buffer_read_object_data_test.cc [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
```
Comment (devbww) [inline google/cloud/storage/internal/grpc_buffer_read_object_data_test.cc:65]: Opt: Ditto.

File: google/cloud/storage/internal/grpc_buffer_read_object_data_test.cc (hunk 9 above)
Comment (coryan) [reply google/cloud/storage/internal/grpc_buffer_read_object_data_test.cc]: Done

File: google/cloud/storage/internal/grpc_buffer_read_object_data.h (hunk 8 above)
Comment (coryan) [reply google/cloud/storage/internal/grpc_buffer_read_object_data.h]: Removed

File: google/cloud/storage/internal/grpc_buffer_read_object_data.cc (hunk 6 above)
Comment (coryan) [reply google/cloud/storage/internal/grpc_buffer_read_object_data.cc]: Done

File: google/cloud/storage/internal/grpc_buffer_read_object_data.h (hunk 5 above)
Comment (coryan) [reply google/cloud/storage/internal/grpc_buffer_read_object_data.h]: Fixed

File: google/cloud/storage/internal/grpc_buffer_read_object_data.h (hunk 4 above)
Comment (coryan) [reply google/cloud/storage/internal/grpc_buffer_read_object_data.h]: Fixed

File: google/cloud/storage/internal/grpc_buffer_read_object_data.h (hunk 3 above)
Comment (coryan) [reply google/cloud/storage/internal/grpc_buffer_read_object_data.h]: Fixed

File: google/cloud/storage/internal/grpc_buffer_read_object_data.h (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/internal/grpc_buffer_read_object_data.h]: Fixed

File: google/cloud/storage/internal/grpc_buffer_read_object_data.h (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/internal/grpc_buffer_read_object_data.h]: Done

Comment (coryan) [reply google/cloud/storage/internal/grpc_buffer_read_object_data.h]: Done

**Pull request merged**